docs: Add getting started page (#193)
Substra/substra-documentation: docs/source/operations/getting-started.rst

***************
Getting started
***************

> This page gives you direction how to run locally the substra stack. This deployment is made of:

Reviewer: ```suggestion
This page gives you directions to locally run the substra stack. This deployment is made of:
```

> * 1 orchestrator (running in standalone mode, runnign its own local DB)

Reviewer: ```suggestion
* 1 orchestrator (running in standalone mode i.e. storing data in its own local database)
```

* 2 backends (running in two organisations, ``org-1`` and ``org-2``)
* 1 frontend

It allows you to run the examples and start using substra SDK (also known as substra).

Prerequisites
=============

Hardware
--------

> The following table indicates the resources needed to run the Substra stack locally. The minimal profile represents resources needed to be able to run the stack, whereas the recommended profile describes how much is needed to have a comfortable development experience.

Reviewer: We need to choose if we write "Substra" (capitalized) or "substra"

.. list-table::
   :header-rows: 1
   :stub-columns: 1

   * -

>      - CPU

Reviewer 2: Nit

```suggestion
     - CPU cores
```

>      - Hard drive space (Go)

Reviewer 2: Nit: the table would be more readable if the units (GB) were in each table cell instead of the header (e.g. 35 GB)

>      - Hard drive space (Go)

Reviewer 2: :fr: 
```suggestion
     - Hard drive space (GB)
```

     - RAM
   * - Minimal
     - 2
     - 35
     - 4
   * - Recommended
     - 4-8
     - 50

>      - 8

Reviewer 2: Is this sufficient? My understanding is we were having problems with only 8 GB RAM

.. caution::
   Choose wisely the parameters passed to Kubernetes as it might try to use all the allocated resources without regards for your system.

Software
--------

* `git <https://git-scm.com/downloads>`_
* `Docker <https://docs.docker.com/>`_ (>= 4.0.0)
*  sed

   .. caution::
      On MacOS you need `gsed`.

* k3d/k3s (>= 5.0.0)
* `kubectl <https://kubernetes.io/>`_
* `skaffold <https://skaffold.dev/>`_
* `helm 3 <https://helm.sh/>`_ (>= 3.7.0)
*  `nodeJS <https://nodejs.org/>`_ (== 16.13.0)

Instructions for Mac
^^^^^^^^^^^^^^^^^^^^

First, install `Homebrew <https://brew.sh/>`_, then run the following commands:

.. code-block:: bash

   brew install k3d
   brew install kubectl
   brew install skaffold
   brew install helm
   brew install gsed # Needed for k3s-create.sh

First time configuration
========================

> 1. Execute the script :download:`k3-create.sh<./getting-started/k3-create.sh>`. This script deletes the existing cluster, recreates a new one and apply a patch for SSL.

Reviewer: ```suggestion
1. Execute the script :download:`k3-create.sh<./getting-started/k3-create.sh>`. This script deletes the existing cluster, recreates a new one and applies a patch for SSL.
```

   1. Download :download:`k3-create.sh<./getting-started/k3-create.sh>`.
   2. Make the script executable.

      .. code-block:: bash

         chmod +x ./k3-create.sh

>    1. Run the script

Author: You're right, thanks for the feedback

      .. code-block:: bash

         ./k3-create.sh

   .. tip::
      This script can be used to reset your development environment.

2. Add the following line to ``/etc/hosts`` to allow the communication between your local cluster and the host (your machine):

   .. code-block:: text

      127.0.0.1 orchestrator.org-1.com orchestrator.org-2.com substra-frontend.org-1.com substra-frontend.org-2.com substra-backend.org-1.com substra-backend.org-2.com

> 3. Add the helm repositories

Reviewer: ```suggestion
3. Add the Helm repositories
```

Author: So should I capitalize all references to Helm? If so, should I apply it to other softwares (e.g. Docker, Skaffold)

Reviewer: If the official name is capitalized, I am in favor of having it capitalized in the doc.

Author: Fair enough, Docker also mentions that the name has to be capitalised.

   .. code-block:: bash

      helm repo add bitnami https://charts.bitnami.com/bitnami
      helm repo add stable https://charts.helm.sh/stable
      helm repo add twuni https://helm.twun.io
      helm repo add jetstack https://charts.jetstack.io

> 4. Clone the various Substra repositories

Reviewer: ```suggestion
4. Clone the Substra component repositories
```

   * `substra <https://github.com/substra/substra>`_

     .. code-block:: bash

      git clone https://github.com/Substra/substra.git

   * `orchestrator <https://github.com/substra/orchestrator>`_

     .. code-block:: bash

      git clone https://github.com/Substra/orchestrator.git

   * `substra-backend <https://github.com/substra/substra-backend>`_

     .. code-block:: bash

      git clone https://github.com/Substra/substra-backend.git

   * `substra-frontend <https://github.com/substra/substra-frontend>`_

     .. code-block:: bash

      git clone https://github.com/Substra/substra-frontend.git

> 5. Install substra in editable mode

Author: I agree with you, and I think this conversation should be carried on outside of this PR as this kind of decision is out-of-scope of this PR. I think some discutions started on this topic inside the organisation. 

For the specific point you mention, I can replace it by Substra SDK, if you think it make it clearer!

> 5. Install substra in editable mode

Reviewer 2: can you guys please get profile pictures? :joy: I can't easily spot who's who

   .. code-block:: bash

      cd substra
      pip install -e .

6. Install frontend dependencies

   .. code-block:: bash

      cd substra-frontend
      npm install --dev

Launching
=========

* Deploy the orchestrator

  .. code-block:: bash

   cd orchestrator
   skaffold run

.. _Deploy the backend:

* Deploy the backend

  .. code-block:: bash

   cd substra-backend
   skaffold run

  .. caution::
     On arm64 architecture (e.g. Apple silicon chips M1 & M2), you need to add the profiles ``dev``and ``arm64``.

     .. code-block:: bash

      skaffold run -p dev,arm64

.. tip::
   When re-launching the orchestrator and the backend, you can speed up the processing by avoiding the update of the chart dependencies using the profile ``nodeps``.

   .. code-block:: bash

      skaffold run -p nodeps

> * Deploy the frontend. You can use two methods (described below)

Reviewer 2: You can also use skaffold run. Feel free to ignore my comment if you intentionally omitted it.

Author: I am not sure to understand this comment - this block is about speeding up the relaunch of orchestrator and backend by not updating chart dependencies.

Reviewer 2: The line I highlighted is

> Deploy the frontend. You can use two methods (described below)

This is independent from orchestrator/backend relaunch?

Reviewer 2: Ok that makes sense if it's the method that is documented :+1: 

Maybe it could be documented on the substra-frontend repo that the frontend can be started with `skaffold run` (that's what I use). cc @HamdyD. But it's out of scope for this PR.

  b. Docker:

     .. code-block:: bash

      docker stop DOCKER_FRONTEND_CONTAINER_NAME

     | with ``DOCKER_FRONTEND_CONTAINER_NAME`` the name of the frontend container you chose during the launch
* Stop the orchestrator:

  .. code-block:: bash

   cd orchestrator
   skaffold delete

* Stop the backend:

  .. code-block:: bash

   cd substra-backend
   skaffold delete

If this command fails and you still have pods up, you can use the following command to remove the ``org-1`` and ``org-2`` namespaces entirely.

.. code-block:: bash

   kubectl rm ns org-1 org-2

Next steps
==========

Now you are ready to go, you are ready to run either the :doc:`/auto_examples/index` or the :doc:`Substrafl (low-level library) examples </substrafl_doc/examples/index>` (low-level library).

If you are interested in more deployment options or more customised set-up, you can have a look at :doc:`/operations/deploy` or at the documentation included in the repo of substra_, substra-backend_, orchestrator_ or substra-frontend_.

Troubleshooting
===============

.. note::
   Before going further in these section, you should check the following points:

>     * Check the version of Skaffold, Helm and Docker. For example, Skaffold is released very often and sometime it introduces bugs, creating unexpected errors.

Reviewer 2: The example is helpful, but in this specific case, how can the user know which skaffold version they should use?

Author: That is an interesting point. Do we have recommendation on Skaffold versions to use already available?

Reviewer 2: Not that I know of :( Usually when there's an unexpected breaking change in skaffold:

- upgrade skaffold
- it doesn't work anymore
- downgrade skaffold to a previous version until it works again

So I guess your sentence makes sense, skaffold does sometimes create unexpected errors. I'll resolve this comment.

    * Check the version of the different Substra components:

      * if you are using a release you can use :ref:`the compatibility table <additional/release:Compatibility table>`.

>       * if you are using the ``latest`` from main, check that you are up-to-date and see if there were any open issue in the repositories or any bugfixes in the latest commits.

Reviewer 2: Nit: it might not be obvious to some uses what "main" refers to here?

```suggestion
      * if you are using the latest commit from the ``main`` git branch, check that you are up-to-date and see if there were any open issue in the repositories or any bugfixes in the latest commits.
```

Author: Good point!

   You can also go through :doc:`the instructions one more time </operations/getting-started>`, maybe they changed since you last saw them.

Troubleshooting prerequisites
-----------------------------

The errors in this category are linked with not reaching the hardware requirements. Please check if `you match these <#hardware>`__ first.

* .. code-block:: pycon

   <ERROR:substra.sdk.backends.remote.rest_client:Requests error status 502: <html>
   <head><title>502 Bad Gateway</title></head>
   <body>
   <center><h1>502 Bad Gateway</h1></center>
   <hr><center>nginx</center>
   </body>
   </html>

   WARNING:root:Function _request failed: retrying in 1s>

>    You may have to increase the number of CPU for the backend in ``substra-backend/charts/substra-backend/values.yaml``

Reviewer 2: I'm not sure what you're referring to here, is it the `resources.requests.cpu` key? However, we don't set those in our chart. Maybe you meant increase the number of CPU you assign to Docker for Mac?

(same below with RAM)

Author: I re-used part of the documentation available, My understanding indeed is that https://github.com/Substra/substra-backend/blob/c9181ce00d9093037e109507bcbcb3e2139d2f27/charts/substra-backend/values.yaml#L173-L182 only applies to deployment and we can use Docker limits for Docker on Mac

Reviewer 2: Ok, so in this section we should probably suggest to increase the resources assigned to Docker for Mac, since changing the values won't help?

* .. code-block:: go

   Unable to connect to the server: net/http: request canceled (Client.Timeout exceeded while awaiting headers)

  .. code-block:: go

   Unable to connect to the server: net/http: TLS handshake timeout

  You may have to increase the RAM for the backend in ``substra-backend/charts/substra-backend/values.yaml``

* If you've got a task with ``FAILED`` status and the logs in the worker are of this form:

  .. code-block:: py3

   substrapp.exceptions.PodReadinessTimeoutError: Pod substra.ai/pod-name=substra-***-compute-*** failed to reach the \"Running\" phase after 300 seconds."

  Your Docker disk image might be full, increase it or clean it with ``docker system prune -a``

Troubleshooting deployment
--------------------------

Skaffold version 1.31.0
^^^^^^^^^^^^^^^^^^^^^^^

> Status check is broken in version 1.31.0 and kubectl secret manifests are not apply until helm deploy is done, but helm deploy depends on kubectl secret manifests.

Reviewer 2: ```suggestion
Status check is broken in version 1.31.0 and kubectl secret manifests are not applied until helm deploy is done, but helm deploy depends on kubectl secret manifests.
```

   failed to deploy: install: exit status 1

As a workaround, you can delete the failing webhook by launching the following command:

.. code-block:: bash

   kubectl delete Validatingwebhookconfigurations ingress-nginx-admission

You should now be able to :ref:`deploy again the backend(s)<Deploy the backend>`.

Other errors during backend deployment
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

If you encounter one of the following errors while deploying the backend:

.. code-block:: bash

   Error: UPGRADE FAILED: cannot patch "orchestrator-org-1-server" with kind Certificate: Internal error occurred: failed calling webhook "webhook.cert-manager.io": Post "https://cert-manager-webhook.cert-manager.svc:443/mutate?timeout=10s": dial tcp <ip>:443: connect: connection refused
   deploying "orchestrator-org-1": install: exit status 1

.. code-block:: bash

   Error from server (InternalError): error when creating "STDIN": Internal error occurred: failed calling webhook "webhook.cert-manager.io": Post "https://cert-manager-webhook.cert-manager.svc:443/mutate?timeout=10s": x509: certificate signed by unknown authority

> Check that the orchestrator is deployed and relaunch the command ``skaffold run``.

Reviewer 2: nevermind, it's correct!

Troubleshooting
===============

.. note::
   Before going further in these section, you should check the following points:
    * Check the version of Skaffold, Helm and Docker. For example, Skaffold is released very often and sometime it introduces bugs, creating unexpected errors.
    * Check the version of the different Substra components:

      * if you are using a release you can use :ref:`the compatibility table <additional/release:Compatibility table>`.
      * if you are using the latest commit from the ``main`` git branch, check that you are up-to-date and see if there were any open issue in the repositories or any bugfixes in the latest commits.

   You can also go through :doc:`the instructions one more time </operations/getting-started>`, maybe they changed since you last saw them.

Troubleshooting prerequisites
-----------------------------

This section summarize errors happening when you are not meeting the hardware requirements. Please check if `you match these <#hardware>`__ first.

.. note::

>    The instructions are targeted to some specific platforms (Docker for Windows in certain cases and Docker for Mac), where you can set the resources allowed to Docker in the configuration panel (information available `here for Mac <https://docs.docker.com/desktop/settings/mac/>`__ and `here for Windows <https://docs.docker.com/desktop/settings/windows/>`__).

Author: @AurelienGasser , can you review the changes I made? (also L281 &L291)

Reviewer 2: LGTM

* .. code-block:: pycon

   <ERROR:substra.sdk.backends.remote.rest_client:Requests error status 502: <html>
   <head><title>502 Bad Gateway</title></head>
   <body>
   <center><h1>502 Bad Gateway</h1></center>
   <hr><center>nginx</center>
   </body>
   </html>

   WARNING:root:Function _request failed: retrying in 1s>

   You may have to increase the number of CPU available in the settings panel.

* .. code-block:: go

   Unable to connect to the server: net/http: request canceled (Client.Timeout exceeded while awaiting headers)

  .. code-block:: go

   Unable to connect to the server: net/http: TLS handshake timeout

  You may have to increase the RAM available in the settings panel.

* If you've got a task with ``FAILED`` status and the logs in the worker are of this form:

  .. code-block:: py3

   substrapp.exceptions.PodReadinessTimeoutError: Pod substra.ai/pod-name=substra-***-compute-*** failed to reach the \"Running\" phase after 300 seconds."

  Your Docker disk image might be full, increase it or clean it with ``docker system prune -a``

Troubleshooting deployment
--------------------------

Skaffold version 1.31.0
^^^^^^^^^^^^^^^^^^^^^^^

> Status check is broken in version 1.31.0 and kubectl secret manifests are not applied until helm deploy is done, but helm deploy depends on kubectl secret manifests.

Reviewer 3: ```suggestion
Status check is broken in version 1.31.0 and kubectl secret manifests are not applied until ``helm deploy`` is done, but ``helm deploy`` depends on kubectl secret manifests.
```

I don't think that would be particularly clear for a user as `helm deploy` is kind of a hidden step in the Skaffold process and you need to know a bit of the Skaffold internals and our deployment specificities to understand the issue.

It has been fixed in `Skaffold 1.32.0 (PR #6574) <https://github.com/GoogleContainerTools/skaffold/releases/tag/v1.32.0>`__.

The solution for the version 1.31.0 is to add ``--status-check=false`` when running Skaffold:

.. code-block:: bash

   skaffold dev/run/deploy --status-check=false

> Failed calling webhook ``validate.nginx.ingress.kubernetes.io``

Reviewer 3: This should not happen with the latest version on `k3s-install.sh` and/or latest version of the stack.

.. code-block:: bash

   kubectl delete Validatingwebhookconfigurations ingress-nginx-admission

You should now be able to :ref:`deploy the backend(s) again<Deploy the backend>`.

Other errors during backend deployment
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

If you encounter one of the following errors while deploying the backend:

.. code-block:: bash

   Error: UPGRADE FAILED: cannot patch "orchestrator-org-1-server" with kind Certificate: Internal error occurred: failed calling webhook "webhook.cert-manager.io": Post "https://cert-manager-webhook.cert-manager.svc:443/mutate?timeout=10s": dial tcp <ip>:443: connect: connection refused
   deploying "orchestrator-org-1": install: exit status 1

.. code-block:: bash

   Error from server (InternalError): error when creating "STDIN": Internal error occurred: failed calling webhook "webhook.cert-manager.io": Post "https://cert-manager-webhook.cert-manager.svc:443/mutate?timeout=10s": x509: certificate signed by unknown authority

Check that the orchestrator is deployed and relaunch the command ``skaffold run``.

Troubleshooting monitoring
--------------------------

k9s limits on log lines
^^^^^^^^^^^^^^^^^^^^^^^

> By default, k9s limits the log to the last 200 lines. To increase this value, set ``logger.tail`` and ``logger.buffer`` to the desired number (e.g. 5000) in the `k9s config file <https://github.com/derailed/k9s#k9s-configuration>`_.

Reviewer 3: This is really some advanced usage of the tools that can be used and I'm not sure it should be part of our documentation.
